lib/scraper.js

```diff
@@ -53,46 +78,81 @@ Scraper.prototype.getResourceHandler = function getHandler (resource) {
 	var depth = resource.getDepth();
 	var depthGreaterThanMax = self.options.maxDepth && depth >= self.options.maxDepth;
 
-	switch (true) {
-		case depthGreaterThanMax: return _.noop;
-		case resource.isCss(): return loadCss;
-		case resource.isHtml(): return loadHtmlAndCss;
-		default: return _.noop;
+	switch(true) {
+		case depthGreaterThanMax:
+			return _.noop;
+		case resource.isCss():
+			return loadCss;
+		case resource.isHtml():
+			return loadHtmlAndCss;
+		default:
+			return _.noop;
 	}
 };
 
 Scraper.prototype.loadResource = function loadResource (resource) {
 	var self = this;
+	var url = resource.getUrl();
 
-	if(!self.options.urlFilter(resource.url)){
+	if(!self.options.urlFilter(url)) {
 		return Promise.resolve(null);
 	}
 
-	// try to find already loaded
-	var loaded = self.getLoadedResource(resource);
+	var loadedResourcePromise = self.getLoadedResourcePromise(url);
+	if(loadedResourcePromise) {
+		return loadedResourcePromise;
+	}
 
-	if (!loaded) {
-		var url = resource.getUrl();
-		var filename = self.options.filenameGenerator(resource, self.options, self.loadedResources);
-		resource.setFilename(filename);
+	var respondedResourcePromise = self.requestResource(resource);
+
+	loadedResourcePromise = respondedResourcePromise.then(function handleResponse (resource) {
+		return Promise.resolve()
```

Comment:
`Promise.resolve()` is a nice way to start a promise chain. Basically it just creates an empty promise to start the chain with. This has two advantages:

1. You don't have to worry whether the result of the first part in the chain actually returns a promise.
2. The code style is normalized.

In this case it is because of the first reason. Calling resourceHandler does not always return a promise, but since it is now in a promise chain, it does not matter.